numba_dpex/_patches.py

```diff
@@ -172,8 +173,71 @@ def _mk_alloc(
     return out
 
 
+def make_queue(context, builder, arrtype):
+    """Utility function used for allocating a new queue.
+
+    This function will allocates a new queue (e.g. SYCL queue)
+    during LLVM code generation (lowering). Given a target context,
+    builder, array type, returns a LLVM value pointing at a numba-dpex
+    runtime allocated queue.
+
+    Args:
+        context (numba.core.base.BaseContext): Any of the context
+            derived from Numba's BaseContext
+            (e.g. `numba.core.cpu.CPUContext`).
+        builder (llvmlite.ir.builder.IRBuilder): The IR builder
+            from `llvmlite` for code generation.
+        arrtype (numba_dpex.core.types.dpnp_ndarray_type.DpnpNdArray):
+            Any of the array types derived from
+            `numba.core.types.nptypes.Array`,
+            e.g. `numba_dpex.core.types.dpnp_ndarray_type.DpnpNdArray`.
+            Please refer to `numba_dpex.dpnp_iface._intrinsic.alloc_empty_arrayobj()`
+            function for details on how to construct this argument.
+
+    Returns:
+        ret (tuple): A tuple containing `llvmlite.ir.instructions.ExtractValue`,
+            `llvmlite.ir.instructions.CastInstr` and `numba.core.pythonapi.PythonAPI`.
+    """
+
+    pyapi = context.get_python_api(builder)
+    queue_struct_proxy = cgutils.create_struct_proxy(
+        DpctlSyclQueue(arrtype.queue)
+    )(context, builder)
+    queue_struct_ptr = queue_struct_proxy._getpointer()
+    queue_struct_voidptr = builder.bitcast(queue_struct_ptr, cgutils.voidptr_t)
+
+    address = context.get_constant(types.intp, id(arrtype.queue))
+    queue_address_ptr = builder.inttoptr(address, cgutils.voidptr_t)
+
+    dpexrtCtx = dpexrt.DpexRTContext(context)
+    dpexrtCtx.queuestruct_from_python(
+        pyapi, queue_address_ptr, queue_struct_voidptr
+    )
+
+    # errcode = dpexrtCtx.queuestruct_from_python(
+    #     pyapi, queue_address_ptr, queue_struct_voidptr
+    # )
+    # is_error = cgutils.is_not_null(builder, errcode)
+    # # Handle error
+    # with builder.if_then(is_error, likely=False):
+    #     pyapi.err_set_string(
+    #         "_patches.make_queue(): PyExc_TypeError",
+    #         "can't unbox dpctl.SyclQueue from PyObject into a Numba "
+    #         "native value. The object maybe of a different type",
+    #     )
+
+    queue_struct = builder.load(queue_struct_ptr)
+    queue_ref = builder.extract_value(queue_struct, 1)
+
+    ret = (queue_ref, queue_address_ptr, pyapi)
+
+    return ret
+
+
 def _empty_nd_impl(context, builder, arrtype, shapes):
```

Comment:
The implementation is getting a bit convoluted as we are trying to maintain API consistency with Numba and monkey patch the Numba function. I suggest, we convert the function to dpex-specific and move it into `dpnp_iface/_intrinsics` and make it specific for our needs. Rather than returning a `queue_ref` LLVM IR object, it is a better design to pass in the object, so that a caller has full control over where the ownership of the LLVM IR instruction.